ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -286,25 +309,27 @@ impl MultiplicationInputsBatch {
         // update last record
         self.last_record = cmp::max(self.last_record, record_id);
 
-        // panics when record_id is less than first_record
-        let id_within_batch = usize::from(record_id) - usize::from(self.first_record);
-        let block_id = (segment.len() * id_within_batch) >> BIT_ARRAY_SHIFT;
-
         // panics when record_id is too large to fit in, i.e. when it is out of bounds
-        if 256 % segment.len() == 0 {
-            self.insert_segment_small(id_within_batch, block_id, segment);
+        if segment.len() <= 256 {
+            self.insert_segment_small(record_id, segment);
         } else {
-            self.insert_segment_large(block_id, &segment);
+            self.insert_segment_large(record_id, &segment);
         }
     }
 
-    /// insert `segments` for `segments` that divide 256
+    /// insert `segments` for `segments` that are smallere than 256
```

Comment:
```suggestion
    /// Insert a `segment` that is smaller than 256 bits
```

It might be worth a debug assertion to check this precondition (and likewise in `insert_segment_large`).